tensorflow_probability/python/math/special.py

```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
   a = tf.broadcast_to(a, broadcast_shape)
   b = tf.broadcast_to(b, broadcast_shape)
   x = tf.broadcast_to(x, broadcast_shape)
-  return tf.math.betainc(a, b, x)
 
-# Derivative implementation based on
-# [1] R. Boik, J. Robinson-Cox,
-#     Derivatives of the Incomplete Beta Function
-#     https://www.jstatsoft.org/article/view/v003i01/beta.der.pdf
+  return tf.math.betainc(a, b, x)
 
 
-def partial_numerator(a, b, x, i):
-  """Partial numerator used in continued fraction expansion of betainc."""
-  f = b * x / (a * (1 - x))
-  result = (a + b + i - 2.) * (a + i - 1) * (b - i) / (
-      (a + 2 * i - 3.) * tf.math.square(a + 2 * i - 2.) * (a + 2 * i - 1.))
-  result = result * tf.math.square(a * f / b) * (i - 1.)
-  return tf.where(
-      tf.math.equal(i, 1.),
-      a * f * (b - 1.) / (b * (a + 1.)),
-      result)
+def _betainc_nth_partial_numerator(iteration, a, b, x, dtype):
+  """Partial numerator used in the continued fraction for betainc."""
+  zero = tf.constant(0., dtype=dtype)
+  one = tf.constant(1., dtype=dtype)
+  two = tf.constant(2., dtype=dtype)
+
+  iteration_is_even = tf.equal(iteration % two, zero)
+
+  # The nth partial numerator is given here: https://dlmf.nist.gov/8.17#E23
+  def even_partial_numerator():
+    # d_{2m}
+    n = iteration / two
+    a_plus_2n = a + two * n
+    a_plus_2n_minus_one = a_plus_2n - one
+    denominator = a_plus_2n * a_plus_2n_minus_one
+
+    db = n * x / denominator
+    value =  db * (b - n)
+    da = -value * (a_plus_2n + a_plus_2n_minus_one) / denominator
+    gradient = tf.concat([da, db], axis=-1)
+    return value, gradient
+
+  def odd_partial_numerator():
+    # d_{2m+1}
+    n = (iteration - one) / two
+    a_plus_n = a + n
+    a_plus_2n = a_plus_n + n
+    a_plus_2n_plus_one = a_plus_2n + one
+    a_plus_b_plus_n = a_plus_n + b
+    denominator = a_plus_2n * a_plus_2n_plus_one
+
+    db = -a_plus_n * x / denominator
+    value =  db * a_plus_b_plus_n
+    da = -value * ((a_plus_2n + a_plus_2n_plus_one) / denominator) - x * (
+        two * a_plus_n + b) / denominator
+    gradient = tf.concat([da, db], axis=-1)
+    return value, gradient
+
+  # Assume iteration is always greater than 1.
+  return tf.cond(
+      iteration_is_even, even_partial_numerator, odd_partial_numerator)
+
+
+def _betainc_modified_lentz_method(a, b, x, dtype, where):
+  """Returns the continued fraction for betainc by modified Lentz's method."""
+  numpy_dtype = dtype_util.as_numpy_dtype(dtype)
+  one = tf.constant(1., dtype=dtype)
+  eps = tf.constant(np.finfo(numpy_dtype).eps, dtype=dtype)
+  tiny = tf.constant(np.finfo(numpy_dtype).tiny, dtype=dtype)
+
+  # max_iterations and tolerance were taken from Cephes.
+  if numpy_dtype == np.float32:
+    max_iterations = 200
+    tolerance = eps
+  else:
+    max_iterations = 600
+    tolerance = tf.constant(3., dtype=dtype) * eps
+
+  small = tf.sqrt(tiny)
+
+  def continued_fraction_evaluation(should_stop, iteration, values, gradients):
+    c, d, f = values
```

Comment:
Can these be renamed to be more descriptive?

Reply:
Sorry for the non-suggestive names. Here I follow the notation used on the Wikipedia page about the [Lentz algorithm](https://en.wikipedia.org/wiki/Lentz%27s_algorithm). On this page, `f_n` is the convergent `f_n = A_n / B_n = f_{n-1} * C_n * D_n` and the quantities `C_n` and `D_n` are the ratios `C_n = A_n / A_{n-1}` and `D_n = B_{n-1} / B_n`. The partial numerator `A_n` and denominator `B_n` are given by the [fundamental recurrence formulas](https://en.wikipedia.org/wiki/Generalized_continued_fraction#Fundamental_recurrence_formulas). Do you have any suggestions?

Maybe:
`convergent` for `f`
`ratio_numerators` for `c`
`ratio_denominators` for `d`
`dconvergent` for `f_grad`
`dratio_numerators` for `c_grad`
`dratio_denominators` for `d_grad`

Reply:
Those names sound good to me! I think having a comment which specifies that map (in case someone is familiar with Lentz's algorithm) would also be welcome (so using the more descriptive names, but saying that ratio_numerators = C_n in a comment.